File: app/src/main/java/com/salino/sali - Copy/data/repository/RecurringRepositoryImpl.kt
package com.salino.sali.data.repository

import com.google.firebase.Timestamp
import com.google.firebase.firestore.ListenerRegistration
import com.salino.sali.data.local.entity.SyncTargetType
import com.salino.sali.data.local.source.ActivityLocalDataSource
import com.salino.sali.data.local.source.RecurringLocalDataSource
import com.salino.sali.data.model.ActivityLog
import com.salino.sali.data.model.ActivityType
import com.salino.sali.data.model.RecurringItem
import com.salino.sali.data.remote.source.RecurringRemoteDataSource
import com.salino.sali.data.sync.SyncQueueProcessor
import com.salino.sali.di.IoDispatcher
import com.salino.sali.domain.repository.AuthRepository
import com.salino.sali.domain.repository.RecurringRepository
import com.salino.sali.util.normalizeItemName
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch
import java.util.Date
import java.util.UUID
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class RecurringRepositoryImpl @Inject constructor(
    private val recurringLocalDataSource: RecurringLocalDataSource,
    private val recurringRemoteDataSource: RecurringRemoteDataSource,
    private val activityLocalDataSource: ActivityLocalDataSource,
    private val authRepository: AuthRepository,
    private val syncQueueProcessor: SyncQueueProcessor,
    @IoDispatcher ioDispatcher: CoroutineDispatcher
) : RecurringRepository {

    private val scope = CoroutineScope(SupervisorJob() + ioDispatcher)
    private val listeners = mutableMapOf<String, ListenerRegistration>()

    override fun observeRecurringItems(householdId: String): Flow<List<RecurringItem>> {
        ensureRemoteSync(householdId)
        return recurringLocalDataSource.observeRecurringItems(householdId)
    }

    override fun observeDueRecurringItems(householdId: String): Flow<List<RecurringItem>> {
        ensureRemoteSync(householdId)
        return recurringLocalDataSource.observeDueRecurringItems(householdId, System.currentTimeMillis())
    }

    override suspend fun upsertRecurringItem(householdId: String, recurringItem: RecurringItem): Result<String> = runCatching {
        val now = Timestamp.now()
        val normalizedName = normalizeItemName(recurringItem.name)
        val recurringId = recurringItem.id.ifBlank { UUID.randomUUID().toString() }
        val updated = recurringItem.copy(
            id = recurringId,
            householdId = householdId,
            normalizedName = normalizedName,
            createdAt = recurringItem.createdAt ?: now,
            updatedAt = now,
            nextDueAt = recurringItem.nextDueAt ?: Timestamp(Date(System.currentTimeMillis() + recurringItem.intervalDays * DAY_MS))
        )
        recurringLocalDataSource.upsertRecurringItem(updated)
        syncQueueProcessor.enqueueUpsert(householdId, SyncTargetType.RECURRING, recurringId)

        val user = authRepository.observeCurrentUser().first()
        val type = if (recurringItem.id.isBlank()) ActivityType.RECURRING_CREATED else ActivityType.RECURRING_UPDATED
        val activity = ActivityLog(
            id = UUID.randomUUID().toString(),
            householdId = householdId,
            type = type.name,
            itemId = recurringId,
            itemName = updated.name,
            actorUserId = user?.id.orEmpty(),
            actorDisplayName = user?.displayName.orEmpty(),
            createdAt = now
        )
        activityLocalDataSource.upsertActivity(activity)
        syncQueueProcessor.enqueueUpsert(householdId, SyncTargetType.ACTIVITY, activity.id)
        syncQueueProcessor.flush(householdId)
        recurringId
    }

    override suspend fun deleteRecurringItem(householdId: String, recurringItemId: String): Result<Unit> = runCatching {
        recurringLocalDataSource.deleteRecurringItem(householdId, recurringItemId)
        syncQueueProcessor.enqueueDelete(householdId, SyncTargetType.RECURRING, recurringItemId)
        syncQueueProcessor.flush(householdId)
    }

    override suspend fun findByNormalizedName(householdId: String, normalizedName: String): Result<RecurringItem?> = runCatching {
        recurringLocalDataSource.findByNormalizedName(householdId, normalizedName)
    }

    override suspend fun updateNextDueDate(householdId: String, normalizedName: String, completedAtMillis: Long): Result<Unit> = runCatching {
        val current = recurringLocalDataSource.findByNormalizedName(householdId, normalizedName) ?: return@runCatching
        val nextDueAt = Timestamp(Date(completedAtMillis + current.intervalDays * DAY_MS))
        val updated = current.copy(
            lastCompletedAt = Timestamp(Date(completedAtMillis)),
            nextDueAt = nextDueAt,
            updatedAt = Timestamp.now()
        )
        recurringLocalDataSource.upsertRecurringItem(updated)
        syncQueueProcessor.enqueueUpsert(householdId, SyncTargetType.RECURRING, updated.id)
        syncQueueProcessor.flush(householdId)
    }

    override suspend fun flushPendingSync(householdId: String): Result<Unit> = runCatching {
        syncQueueProcessor.flush(householdId)
    }

    private fun ensureRemoteSync(householdId: String) {
        if (listeners.containsKey(householdId)) return
        listeners[householdId] = recurringRemoteDataSource.listenToRecurringItems(householdId) { items ->
            scope.launch {
                recurringLocalDataSource.mergeRemoteRecurringItems(householdId, items)
                syncQueueProcessor.flush(householdId)
            }
        }
    }

    companion object {
        private const val DAY_MS = 24 * 60 * 60 * 1000L
    }
}

File: ios/Data/Repositories/RecurringRepositoryImpl.swift
import Foundation
import FirebaseFirestore

@MainActor
final class RecurringRepositoryImpl: RecurringRepository {
    private let localStore: LocalStore
    private let remoteDataSource: RecurringRemoteDataSource
    private let authRepository: AuthRepository
    private let syncQueueProcessor: SyncQueueProcessor
    private var listeners: [String: ListenerRegistration] = [:]
    private let day: TimeInterval = 24 * 60 * 60

    init(localStore: LocalStore, remoteDataSource: RecurringRemoteDataSource, authRepository: AuthRepository, syncQueueProcessor: SyncQueueProcessor) {
        self.localStore = localStore
        self.remoteDataSource = remoteDataSource
        self.authRepository = authRepository
        self.syncQueueProcessor = syncQueueProcessor
    }

    func observeRecurringItems(householdId: String) -> AsyncStream<[RecurringItem]> {
        ensureRemoteSync(householdId)
        return localStore.observeRecurringItems(householdId: householdId)
    }

    func observeDueRecurringItems(householdId: String) -> AsyncStream<[RecurringItem]> {
        ensureRemoteSync(householdId)
        return localStore.observeRecurringItems(householdId: householdId, dueOnly: true)
    }

    func upsertRecurringItem(householdId: String, recurringItem: RecurringItem) async throws -> String {
        let now = Date()
        var item = recurringItem
        item.id = item.id.isEmpty ? UUID().uuidString : item.id
        item.householdId = householdId
        item.normalizedName = normalizeItemName(item.name)
        item.createdAt = item.createdAt ?? now
        item.updatedAt = now
        item.nextDueAt = item.nextDueAt ?? now.addingTimeInterval(TimeInterval(item.intervalDays) * day)
        localStore.upsertRecurringItem(item)
        syncQueueProcessor.enqueueUpsert(householdId: householdId, targetType: .recurring, targetId: item.id)
        try await logRecurringActivity(householdId: householdId, recurring: item, isNew: recurringItem.id.isEmpty)
        try? await syncQueueProcessor.flush(householdId: householdId)
        return item.id
    }

    func deleteRecurringItem(householdId: String, recurringItemId: String) async throws {
        localStore.deleteRecurringItem(householdId: householdId, recurringItemId: recurringItemId)
        syncQueueProcessor.enqueueDelete(householdId: householdId, targetType: .recurring, targetId: recurringItemId)
        try? await syncQueueProcessor.flush(householdId: householdId)
    }

    func findByNormalizedName(householdId: String, normalizedName: String) async throws -> RecurringItem? {
        localStore.findRecurringItem(householdId: householdId, normalizedName: normalizedName)
    }

    func updateNextDueDate(householdId: String, normalizedName: String, completedAt: Date) async throws {
        guard var current = localStore.findRecurringItem(householdId: householdId, normalizedName: normalizedName) else { return }
        current.lastCompletedAt = completedAt
        current.nextDueAt = completedAt.addingTimeInterval(TimeInterval(current.intervalDays) * day)
        current.updatedAt = Date()
        localStore.upsertRecurringItem(current)
        syncQueueProcessor.enqueueUpsert(householdId: householdId, targetType: .recurring, targetId: current.id)
        try? await syncQueueProcessor.flush(householdId: householdId)
    }

    func flushPendingSync(householdId: String) async throws {
        try await syncQueueProcessor.flush(householdId: householdId)
    }

    func clearListeners() {
        listeners.values.forEach { $0.remove() }
        listeners.removeAll()
    }

    private func ensureRemoteSync(_ householdId: String) {
        guard listeners[householdId] == nil else { return }
        listeners[householdId] = remoteDataSource.listenToRecurringItems(householdId: householdId) { [weak self] items in
            Task { @MainActor in
                guard let self else { return }
                localStore.mergeRemoteRecurringItems(householdId: householdId, items: items)
                try? await syncQueueProcessor.flush(householdId: householdId)
            }
        }
    }

    private func logRecurringActivity(householdId: String, recurring: RecurringItem, isNew: Bool) async throws {
        let user = await firstCurrentUser()
        let activity = ActivityLog(
            id: UUID().uuidString,
            householdId: householdId,
            type: isNew ? .recurringCreated : .recurringUpdated,
            itemId: recurring.id,
            itemName: recurring.name,
            actorUserId: user?.id ?? "",
            actorDisplayName: user?.displayName ?? "",
            createdAt: Date()
        )
        localStore.upsertActivity(activity)
        syncQueueProcessor.enqueueUpsert(householdId: householdId, targetType: .activity, targetId: activity.id)
    }

    private func firstCurrentUser() async -> UserProfile? {
        for await user in authRepository.observeCurrentUser() { return user }
        return nil
    }
}
